kernel/work_q.c

```diff
@@ -147,4 +147,82 @@ bool k_delayed_work_pending(struct k_delayed_work *work)
 	       k_work_pending(&work->work);
 }
 
+static void work_delayable_timeout(struct _timeout *t)
+{
+	struct k_work_delayable *w = CONTAINER_OF(t, struct k_work_delayable,
+						   timeout);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	struct k_work_q *wq = w->work_q;
+
+	w->work_q = NULL;
+	k_spin_unlock(&lock, key);
+
+	/* submit work to workqueue */
+	k_work_submit_to_queue(wq, &w->work);
+}
+
+void k_work_delayable_init(struct k_work_delayable *work, k_work_handler_t handler)
+{
+	k_work_init(&work->work, handler);
+	z_init_timeout(&work->timeout);
+	work->work_q = NULL;
+}
+
+int k_work_schedule_for_queue(struct k_work_q *work_q,
+			      struct k_work_delayable *work,
+			      k_timeout_t delay)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	int err = 0;
+
+	/* Work cannot be active in multiple queues */
+	if (work->work_q != NULL && work->work_q != work_q) {
+		err = -EBUSY;
+		goto unlock;
+	}
+
+	/* Cancel any incomplete timeout. */
+	if (work->work_q == work_q) {
```

Comment:
Writing "work->work_q != NULL" would be clearer here, IMHO.  The point is to test if there is an active timeout.

Reply:
Refactored to be a bit simpler.